### PR TITLE
12 new datasets

### PR DESCRIPTION
**Pull request by AlexandreMisrahi2005**
Configs, prompts, data processors for: 
[query (datastore)]
- bioasq12b (pubmed abstracts)
- covidQA (CORD-19 chunks)
- FiQA (forum datastore)
- MultiQA (kilt100w + ms-marco corpora)
- ParaphraseRC (movie plot chunks)
- RobustQA Lifestyle (LoTTE chunks)
- RobustQA Recreation (LoTTE chunks)
- RobustQA Science (LoTTE chunks)
- RobustQA Technology (LoTTE chunks)
- RobustQA Writing (LoTTE chunks)
- SearchQA (google snippets corpus)
- TechQA (documentation corpus)

all unit tests passed except test_train_lora, waiting upon PR #27 to merge

### CODE REVIEW

File: modules/processors/multidomain_dataset_processor.py [hunk 1]
```diff
@@ -1,15 +1,101 @@
...
+        self.dev_path = dev_path
+        super().__init__(*args, **kwargs, dataset_name=self.dataset_name)
+
+    def process(self):
+        seed = 42
+        if self.split not in ["train", "dev"]:
+            raise ValueError("split should be 'train' or 'dev'")
+        all_data = []
+        with zipfile.ZipFile(self.train_path, 'r') as z:
+            with z.open('BioASQ-training12b/training12b_new.json') as json_file:
```
Comment (sclincha) [inline modules/processors/multidomain_dataset_processor.py:35]: Not sure anyone else can get this dataset? ie How can I get this file? Should we upload it to HF? 
BioASQ-training12b/training12b_new.json'

File: modules/processors/multidomain_dataset_processor.py [hunk 2]
```diff
@@ -444,3 +527,396 @@ def chunk_text(text, title, max_size=1000, overlap=200):
...
+    Other possible source: https://huggingface.co/colbertv2
+    """
+    def __init__(self, path, *args, **kwargs):
+        dataset_name = 'LoTTE'
+        self.path = path
+        super().__init__(*args, **kwargs, dataset_name=dataset_name)
+
+    def process(self):
+        import tarfile
+        # download https://downloads.cs.stanford.edu/nlp/data/colbert/colbertv2/lotte.tar.gz in some directory
```
Comment (sclincha) [inline modules/processors/multidomain_dataset_processor.py:751]: Sorry .... but Can we automate this download? and process the result on the fly?

File: modules/processors/multidomain_dataset_processor.py (hunk 1 above)
Comment (AlexandreMisrahi2005) [reply modules/processors/multidomain_dataset_processor.py]: Done

File: modules/processors/multidomain_dataset_processor.py (hunk 2 above)
Comment (AlexandreMisrahi2005) [reply modules/processors/multidomain_dataset_processor.py]: Done

**Comment by sclincha** [comment]
Looks good to me.  @maxime-louis any feebdack?

File: modules/processors/utils.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
+'''
+Utility functions for dataset processors
+'''
+
+def chunk_text(text, id, title=None, max_size=1000, overlap=200, words_or_chars='chars'):
```
Comment (maxime-louis) [inline modules/processors/utils.py:5]: type hints pls :): text: str, title: str = None etc

File: modules/processors/utils.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        id (str): The id of the text.
+        title (str): The title of the text. If None, no title is prepended.
+        max_size (int): The maximum size of each chunk. Can be for characters or words.
+        overlap (int): The overlap between chunks. Can be for characters or words.
+        words_or_chars (str): Whether to chunk by characters ('chars') or words ('words'). Default is 'chars'.
+
+    Returns:
+        List[Dict[str, str]]: list of chunks and ids.
+    """
+    if title is None:
```
Comment (maxime-louis) [inline modules/processors/utils.py:20]: title = title or "" is a more compact syntax

File: modules/processors/multidomain_dataset_processor.py [hunk 5]
```diff
@@ -444,3 +536,411 @@ def chunk_text(text, title, max_size=1000, overlap=200):
...
+class TechQA_docs(Processor):
+    def __init__(self, *args, **kwargs):
+        dataset_name = 'TechQA_docs'
+        super().__init__(*args, **kwargs, dataset_name=dataset_name)
+
+    def process(self):
+        ds = datasets.load_dataset("rojagtap/tech-qa")
+        dataset = datasets.concatenate_datasets([ds["train"], ds["validation"], ds["test"]])
+
+        def chunk_text(text, title, id, max_size=1000, overlap=200):
```
Comment (maxime-louis) [inline modules/processors/multidomain_dataset_processor.py:597]: Can't you re-use the chunk_text method that you added in utils there ?
Also: type hints :)

File: modules/processors/multidomain_dataset_processor.py [hunk 6]
```diff
@@ -444,3 +536,411 @@ def chunk_text(text, title, max_size=1000, overlap=200):
...
+    """
+    def __init__(self, *args, **kwargs):
+        dataset_name = 'ParaphraseRC'
+        super().__init__(*args, **kwargs, dataset_name=dataset_name)
+
+    def process(self):
+        def map_fn(row):
+            row["content"] = f"{row['title']}: {row['content']}"
+            return row
+        dataset = datasets.load_dataset("ibm/duorc", "ParaphraseRC")[self.split].filter(lambda x: not x["no_answer"]).rename_columns({"question_id":"id", "question":"content", "answers":"label"}).map(map_fn, num_proc=self.num_proc).remove_columns(["plot_id", "plot", "title", "no_answer"])
```
Comment (maxime-louis) [inline modules/processors/multidomain_dataset_processor.py:650]: line too long please split

File: modules/processors/multidomain_dataset_processor.py [hunk 7]
```diff
@@ -444,3 +536,411 @@ def chunk_text(text, title, max_size=1000, overlap=200):
...
+        super().__init__(*args, **kwargs, dataset_name=dataset_name)
+
+    def process(self):
+        dataset = datasets.load_dataset("ibm/duorc", "ParaphraseRC")
+        if self.split == 'all':
+            dataset = datasets.concatenate_datasets([dataset["train"], dataset["validation"], dataset["test"]])
+        else:
+            dataset = dataset[self.split]
+
+        def chunk_text(text, title, id, max_size=1000, overlap=200):
```
Comment (maxime-louis) [inline modules/processors/multidomain_dataset_processor.py:665]: same comment: this looks like duplicated code ?

File: modules/processors/multidomain_dataset_processor.py [hunk 8]
```diff
@@ -444,3 +536,411 @@ def chunk_text(text, title, max_size=1000, overlap=200):
...
+    """
+    def __init__(self, *args, **kwargs):
+        dataset_name = 'CovidQA'
+        super().__init__(*args, **kwargs, dataset_name=dataset_name)
+
+    def process(self):
+        def map_fn(row):
+            row["label"] = row["answers"]["text"]
+            return row
+        dataset = datasets.load_dataset("deepset/covid_qa_deepset")[self.split].rename_column("question","content").map(map_fn, num_proc=self.num_proc).remove_columns(["document_id", "context", "is_impossible", "answers"]).cast_column('id', datasets.Value('string'))
```
Comment (maxime-louis) [inline modules/processors/multidomain_dataset_processor.py:724]: lien too long please split

File: modules/processors/multidomain_dataset_processor.py [hunk 9]
```diff
@@ -444,3 +536,411 @@ def chunk_text(text, title, max_size=1000, overlap=200):
...
+        dataset = datasets.Dataset.from_pandas(pd.DataFrame(all_chunks))
+
+        return dataset
+    
+class LoTTE(Processor):
+    """
+    Source: https://downloads.cs.stanford.edu/nlp/data/colbert/colbertv2/lotte.tar.gz
+    Other possible source: https://huggingface.co/colbertv2
+    """
+    def __init__(self, url, tmp_download_dir='raw_datasets', *args, **kwargs):
```
Comment (maxime-louis) [inline modules/processors/multidomain_dataset_processor.py:753]: so tmp_download_dir is only used once ?
You should use 'tempfile' package instead:
```
import tempfile

# Create a temporary directory
with tempfile.TemporaryDirectory() as temp_dir:
    print(f"Temporary directory created at: {temp_dir}")
    # Use the temporary directory here

# The directory and its contents are automatically cleaned up when the block ends
```

File: modules/processors/utils.py (hunk 3 above)
Comment (AlexandreMisrahi2005) [reply modules/processors/utils.py]: done

File: modules/processors/multidomain_dataset_processor.py (hunk 5 above)
Comment (AlexandreMisrahi2005) [reply modules/processors/multidomain_dataset_processor.py]: done

File: modules/processors/multidomain_dataset_processor.py (hunk 6 above)
Comment (AlexandreMisrahi2005) [reply modules/processors/multidomain_dataset_processor.py]: done

File: modules/processors/multidomain_dataset_processor.py (hunk 7 above)
Comment (AlexandreMisrahi2005) [reply modules/processors/multidomain_dataset_processor.py]: done

File: modules/processors/multidomain_dataset_processor.py (hunk 8 above)
Comment (AlexandreMisrahi2005) [reply modules/processors/multidomain_dataset_processor.py]: done

File: modules/processors/multidomain_dataset_processor.py (hunk 9 above)
Comment (AlexandreMisrahi2005) [reply modules/processors/multidomain_dataset_processor.py]: done, thanks for the tip 👍

**Comment by maxime-louis** [comment]
It looks good to me too now !

**Pull request merged**